### PR TITLE
Fix scatter grad bug

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

Bug fixes
### PR changes

Others
### Describe

When scatter calculates the gradient of X in bp, this is not an inplace operation, which is wrong. To fix the bug here, select the corresponding position by index and set it to 0.

**[op benchmark CI]** During the reverse calculation of scatter, the gradient of x was not calculated correctly. The related repairs were made here, which will cause the reverse calculation of scatter to slow down and meet expectations.

### CODE REVIEW

File: paddle/fluid/operators/scatter.cu.h [hunk 1]
```diff
@@ -138,6 +137,30 @@ void GPUScatterAssign(const framework::ExecutionContext& context,
...
+  int slice_size = 1;
+  for (int i = 1; i < dst_dims.size(); ++i) slice_size *= dst_dims[i];
+  const IndexT* p_index = index.data<IndexT>();
+  T* p_output = output->data<T>();
+  const size_t& slice_bytes = slice_size * sizeof(T);
+
+  // set block and grid num
+  int block = 512;
+  int n = slice_size * index_size;
+  int grid = (n + block - 1) / block;
```
Comment (wawltor) [inline paddle/fluid/operators/scatter.cu.h]: 这块的grid有最大的限制，可以根据GetCUDAMaxGridDimSize来做一个限制

File: paddle/fluid/operators/scatter.h [hunk 2]
```diff
@@ -171,6 +171,24 @@ void ScatterAssignAdd(const framework::ExecutionContext& ctx, const Tensor& src,
...
+  T* p_output = output->data<T>();
+  size_t slice_size = 1;
+  for (int i = 1; i < dst_dims.size(); ++i) slice_size *= dst_dims[i];
+  const size_t slice_bytes = slice_size * sizeof(T);
+  for (int i = 0; i < index_size; ++i) {
+    const IndexT& index_ = p_index[i];
+    memset(p_output + slice_size * index_, 0, slice_bytes);
+  }
+}
+
```
Comment (wawltor) [inline paddle/fluid/operators/scatter.h:191]: scatter_nd和scatter很相似，scatter_nd是否有同样的问题？

File: paddle/fluid/operators/scatter.cu.h [hunk 3]
```diff
@@ -138,6 +137,30 @@ void GPUScatterAssign(const framework::ExecutionContext& context,
       p_src, p_index, p_output, index_size, slice_size, overwrite);
 }
 
+// The function is only for scatter grad x,
+// however update grad use gather
+template <typename T, typename IndexT = int>
+void GPUScatterGradForX(const platform::DeviceContext& ctx, const Tensor& index,
+                        Tensor* output) {
+  int index_size = index.dims()[0];
```
Comment (wawltor) [inline paddle/fluid/operators/scatter.cu.h]: int可能会遇到越界的问题，用IndexT会更好一点，下面看起来都有一些的int32、int64的问题，可能需要看一下

File: paddle/fluid/operators/scatter.cu.h (hunk 1 above)
Comment (ForFishes) [reply paddle/fluid/operators/scatter.cu.h]: ok，done

File: paddle/fluid/operators/scatter.h (hunk 2 above)
Comment (ForFishes) [reply paddle/fluid/operators/scatter.h:191]: 没有这个问题，原因是scatter_nd其实是scatter_nd_add，采用的保留原来值，将update加上去，因此梯度可以直接赋值。而scatter的逻辑是会将原来位置的值给置为0，导致反向传播的梯度为0。

File: paddle/fluid/operators/scatter.cu.h (hunk 3 above)
Comment (ForFishes) [reply paddle/fluid/operators/scatter.cu.h]: 好的。我一并修改下。

File: paddle/fluid/operators/scatter.h (hunk 2 above)
Comment (iclementine) [reply paddle/fluid/operators/scatter.h:191]: 其实这里也可以直接用  scatter 来实现。
dx 就是把一个和 updates 一样形状的 0 scatter 到 dout 上。

**Review by iclementine — APPROVED** [review]
LGTM！

**Review by GaoWei8 — APPROVED** [review]
LGTM for op benchmark ci

**Review by wawltor — APPROVED** [review]
LGTM

**Pull request merged**